File: lager2/katthalsbandet/submissions/partially_accepted/claude_n3.cpp
// @EXPECTED_GRADES@ TLE AC AC TLE TLE TLE TLE TLE
// O(N^3) DP: for each rotation, O(N^2) DP over (position, balance)
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

int main() {
    int n;
    ll a, b;
    scanf("%d %lld %lld", &n, &a, &b);
    char s[500001];
    scanf("%s", s);

    const ll INF = 1e18;
    ll best = INF;
    vector<ll> dp(n + 1), ndp(n + 1);

    for (int c = 0; c < n; c++) {
        fill(dp.begin(), dp.begin() + n + 1, INF);
        dp[0] = 0;
        int lo = 0, hi = 0;

        for (int idx = 0; idx < n; idx++) {
            int pos = (c + idx) % n;
            ll keep = (pos < c) ? b : 0;
            bool open = s[pos] == '(';

            int nlo = n, nhi = -1;
            // Only clear the range we might write to
            for (int j = max(0, lo - 1); j <= min(n, hi + 1); j++)
                ndp[j] = INF;

            for (int j = lo; j <= hi; j++) {
                if (dp[j] >= INF) continue;
                // Remove
                ll v = dp[j] + a;
                if (v < ndp[j]) {
                    ndp[j] = v;
                    if (j < nlo) nlo = j;
                    if (j > nhi) nhi = j;
                }
                // Keep
                int nj = open ? j + 1 : j - 1;
                if (nj >= 0 && nj <= n) {
                    v = dp[j] + keep;
                    if (v < ndp[nj]) {
                        ndp[nj] = v;
                        if (nj < nlo) nlo = nj;
                        if (nj > nhi) nhi = nj;
                    }
                }
            }

            swap(dp, ndp);
            lo = nlo;
            hi = nhi;
        }

        best = min(best, dp[0]);
    }

    printf("%lld\n", best);
}

File: lager2/katthalsbandet/submissions/partially_accepted/claude_n3.py
# @EXPECTED_GRADES@ TLE AC AC TLE TLE TLE TLE TLE
# O(N^3) DP: for each rotation, O(N^2) DP over (position, balance)

def solve():
    n, a, b = map(int, input().split())
    s = input()

    INF = float('inf')
    best = INF

    for c in range(n):
        dp = [INF] * (n + 1)
        dp[0] = 0
        lo = hi = 0

        for idx in range(n):
            pos = (c + idx) % n
            keep_cost = b if pos < c else 0
            is_open = s[pos] == '('

            ndp = [INF] * (n + 1)
            nlo, nhi = n, -1

            for j in range(lo, hi + 1):
                if dp[j] >= INF:
                    continue
                # Remove this character
                v = dp[j] + a
                if v < ndp[j]:
                    ndp[j] = v
                    if j < nlo: nlo = j
                    if j > nhi: nhi = j
                # Keep this character
                nj = j + 1 if is_open else j - 1
                if 0 <= nj <= n:
                    v = dp[j] + keep_cost
                    if v < ndp[nj]:
                        ndp[nj] = v
                        if nj < nlo: nlo = nj
                        if nj > nhi: nhi = nj

            dp = ndp
            lo, hi = nlo, nhi

        best = min(best, dp[0])

    print(best)

solve()
